### PR TITLE
Chrome: Adding the "add new term form" for hierarchical taxonomies

### LINKED ISSUE #2233: Add new category option
0.7 version. Using Twenty Seventeen. Local setup in Desktop Server. In Chrome.

Right now I can not see a way to add a new Category.

Here is what it looks like in Gutenberg right now.
screen shot 2017-08-04 at 23 53 58

Here is what is looks like in WordPress today.
A few more options in place.

screen shot 2017-08-04 at 23 54 39

There are a few issues in regards to Category but I did not notice one that brought up this issue.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #2233

screen shot 2017-08-21 at 15 33 50

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice! Looks like it's working as intended.

I was a bit confused by the usage of the word "Term" instead of "Category", though — is there a good reason for that?

**Comment by youknowriad** [comment]
@jasmussen Yeah, the reason is this form can be used for custom taxonomies too. We have the same issue for tags. @aduth suggested adding a special case for Tags and Categories, do you think we should do it?

**Comment by jasmussen** [comment]
Yeah if we could spell out "category" for the taxonomy WordPress is born with, then it would be totally cool to use "term" for and custom hierarchical taxonomy.

**Comment by TimothyBJacobs** [comment]
Taxonomies are registered with labels for custom taxonomies. Can that not be used?

**Comment by aduth** [comment]
>Taxonomies are registered with labels for custom taxonomies. Can that not be used?

This would be good, but from what I can tell these labels are not exposed through the REST API? The only option we'd have is to bootstrap them into the client server-side (localize), which we've shied away from as it should signal that the REST API is missing necessary functionality.

File: editor/sidebar/post-taxonomies/hierarchical-term-selector.js [hunk 1]
```diff
@@ -48,16 +58,87 @@ class HierarchicalTermSelector extends Component {
...
+			// Tries to create a term or fetch it if it already exists
+			const Model = wp.api.getTaxonomyModel( this.props.slug );
+			this.addRequest = new Model( {
+				name: formName,
+				parent: formParent ? formParent : undefined,
+			} ).save();
+			this.addRequest
+				.then( resolve, ( xhr ) => {
+					const errorCode = xhr.responseJSON && xhr.responseJSON.code;
+					if ( errorCode === 'term_exists' ) {
```
Comment (timmyc) [inline editor/sidebar/post-taxonomies/hierarchical-term-selector.js]: Any chance we can do a client-side validation against known terms to help bypass this?

**Review by timmyc — COMMENTED** [review]
This is like a trip down memory lane 😄  - so I hit a few issues while testing the branch out.  While adding a term that has the same name, and parent of one that already exists, the resulting API request fails, and the term is shown in the UI, along with console errors of duplicate keys:

videos___gutenberg_ _wordpress_develop_ _wordpress

videos___gutenberg_ _wordpress_develop_ _wordpress 2

Upon hard-refreshing the page, the errant duplicate named category is not shown.

The parent drop-down also shows terms at all levels without indentation.  So if I have a term name of test at 2 different levels in different "trees", it looks a bit confusing:

videos___gutenberg_ _wordpress_develop_ _wordpress

Here is the same form control in the tinymce chrome:

edit_post_ _wordpress_develop_ _wordpress

File: editor/sidebar/post-taxonomies/hierarchical-term-selector.js (hunk 1 above)
Comment (youknowriad) [inline editor/sidebar/post-taxonomies/hierarchical-term-selector.js]: We have the same issue with tags, the current state of the API doesn't allow another approach.

**Comment by youknowriad** [comment]
> This is like a trip down memory lane

@timmyc Indeed, Gutenberg feels like that in several places :)

--
So, the failing request is still there, but it's being handled correctly now. And I fixed the selector indentation.

**Comment by paulwilde** [comment]
It would be nice if there was an advanced mode (Maybe this could be an opt-in feature per taxonomy?) where when you add a new taxonomy you can also add additional fields such as the description and slug. Maybe also the ability to update/remove existing taxonomies thus eliminating the need to move out from this screen to manage your taxonomies and make it more of a single-app experience.

This is obviously more of a version 2.0 thing though and I don't want to derail this topic in anyway. Just spitting out ideas.

File: editor/sidebar/post-taxonomies/hierarchical-term-selector.js [hunk 2]
```diff
@@ -109,16 +182,63 @@ class HierarchicalTermSelector extends Component {
 		} );
 	}
 
+	renderParentSelectorOptions( terms, level = 0 ) {
```
Comment (timmyc) [inline editor/sidebar/post-taxonomies/hierarchical-term-selector.js]: <3 this is some pretty recursion here 🙌

**Review by timmyc — APPROVED** [review]
All aforementioned issues fixed up. Nice side-affect of attempting to add an pre-existing term is the existing one gets checked - nice touch!

**Comment by aduth** [comment]
Noted at https://github.com/WordPress/gutenberg/issues/2499#issuecomment-324339573 , taxonomy labels _are_ exposed through the REST API currently, so we could enhance these components to use the labels.

**Pull request merged**